### ISSUE TITLE
Sandbox OAuth url overwritten with production url after retry

### ISSUE THREAD

**Issue by wozniak-michal** [forcedotcom/dataloader#1245]
**Describe the bug**
The same issue as described here: https://github.com/forcedotcom/dataloader/issues/1020
When attempting to login to a Sandbox with OAuth, the URL in use is login.salesforce.com instead of test.salesforce.com. This prevents the login from proceeding successfully.

**To Reproduce**
1. Turn off internet connection (turn off wifi connection, unplug ethernet cable)
2. Open Data Loader 62.0.1
3. Click any action, e.g. Insert
4. Click OAuth radio
5. Choose Sandbox from the Environment dropdown
6. Click Log in
7. Close data loader

**Expected behavior**
- `util.OAuthBrowserLoginRunner retryBrowserLoginWithDefaultURL` should retry with `test.salesforce.com` OAuth server URL
-  `sfdc.oauth.Sandbox.server` value in `config.properties` file shouldn't be overwritten with `login.salesforce.com`

**Screenshots**
```
2024-08-29 20:25:24,835 INFO  [main] action.OperationInfoUIHelper instantiateWizard (OperationInfoUIHelper.java:113) - Instantiating wizard: insert
2024-08-29 20:25:43,792 WARN  [main] util.OAuthBrowserLoginRunner <init> (OAuthBrowserLoginRunner.java:77) - Failed to authenticate using OAuth server URL https://test.salesforce.com. Error: No route to host: connect
2024-08-29 20:25:43,792 INFO  [main] util.OAuthBrowserLoginRunner retryBrowserLoginWithDefaultURL (OAuthBrowserLoginRunner.java:97) - Retrying to authenticate using OAuth server URL https://login.salesforce.com
2024-08-29 20:25:43,795 ERROR [main] ui.OAuthLoginFromBrowserFlow open (OAuthLoginFromBrowserFlow.java:78) - No such host is known (login.salesforce.com)
2024-08-29 20:25:45,174 ERROR [Thread-2] util.OAuthBrowserLoginRunner$1 run (OAuthBrowserLoginRunner.java:223) - No route to host: connect
```

**Desktop (please complete the following information):**
 - OS: Windows 11 and Mac OS Ventura
 - Dataloader Version: 62.0.1

**Additional context**

**Comment by ashitsalesforce** [comment]
Hi @wozniak-michal, thanks for pointing out this issue. the workaround for now is to set the following properties in your `config.properties` file:
```
sfdc.oauth.Sandbox.server=https\://<your sandbox server name>.salesforce.com/
sfdc.oauth.Sandbox.redirecturi=https\://<your sandbox server name>.salesforce.com/services/oauth2/success
```

I am attaching a potential fix containing the following logic:
If `sfdc.oauth.Sandbox.server` is set to `https\://test.salesforce.com/`, the code uses the value of `sfdc.endpoint` property. Similar logic for `sfdc.oauth.Production.server`, `sfdc.oauth.Sandbox.redirecturi`, `sfdc.oauth.Production.redirecturi` properties.

[dataloader_v62.0.1.zip](https://github.com/user-attachments/files/16823473/dataloader_v62.0.1.zip)

MD5 (dataloader_v62.0.1.zip) = 4fa82e07d1df7911ff265d891142db34

**Comment by wozniak-michal** [comment]
Hi @ashitsalesforce, thanks for the response!

> Are you saying that if sfdc.oauth.Sandbox.server is set to test.salesforce.com and that fails, the code should retry with the same URL instead of default endpoint (login.salesforce.com)?

Correct! If user selected Sandbox Environment it should never try to use `login.salesforce.com`. It should always be either custom domain (provided by user) or default `test.salesforce.com` endpoint.

> I am also attaching a potential fix containing the following logic:
If sfdc.oauth.Sandbox.server is set to https\://test.salesforce.com/, the code uses the value of sfdc.endpoint property. Similar logic implemented for sfdc.oauth.Production.server, sfdc.oauth.Sandbox.redirecturi, sfdc.oauth.Production.redirecturi properties. If this URL does not work (does not return OAuth handshake response), default endpoint (login.salesforce.com) is used to attempt OAuth login.

I don't think the behaviour from the last sentence make sense. Default endpoint should be determined by the selected environment, and should never change by itself from sandbox to production and vice versa.

**Comment by ashitsalesforce** [comment]
Hi @wozniak-michal , thanks for clarifying these questions. Here is the outline of the fix (to be implemented) based on your responses and legacy code:

- transition `sfdc.endpoint` into two properties for - `sfdc.endpoint.production` and `sfdc.endpoint.sandbox`
- default for `sfdc.endpoint.production` will be `https://login.salesforce.com` and default for `sfdc.endpoint.sandbox` will be `https://test.salesforce.com`
- if any of the 4 legacy OAuth properties for `server` and `redirecturi` are set to a value other than the default, use them instead of using `sfdc.endpoint.production` or `sfdc.endpoint.sandbox` values.

Let us know if that will align with the expected behavior.

**Comment by wozniak-michal** [comment]
@ashitsalesforce makes more sense I think! At least users won't be left with a broken config.

**Comment by ashitsalesforce** [comment]
PRs to fix the issue: #1246 , #1247 , #1249 , #1250
